hibernate-core/src/main/java/org/hibernate/action/internal/EntityInsertAction.java

```diff
@@ -150,6 +153,168 @@ public void execute() throws HibernateException {
 
 		handleNaturalIdPostSaveNotifications( id );
 
+		EntityMetamodel entityMetamodel = getPersister().getEntityMetamodel();
+
+		if (entityMetamodel.getBytecodeEnhancementMetadata().isEnhancedForLazyLoading()) {
+			PersistentAttributeInterceptable interceptable = ((PersistentAttributeInterceptable) instance );
+			PersistentAttributeInterceptor interceptor = interceptable.$$_hibernate_getInterceptor();
+			final BytecodeEnhancementMetadata enhancementMetadata = entityMetamodel.getBytecodeEnhancementMetadata();
+			final LazyAttributesMetadata lazyAttributesMetadata = enhancementMetadata.getLazyAttributesMetadata();
+			if (interceptor == null) {
+				Type[] types = persister.getPropertyTypes();
+				final String[] propertyNames = persister.getPropertyNames();
+				final CascadeStyle[] cascadeStyles = persister.getPropertyCascadeStyles();
+				boolean needsInterceptor = false;
+				Set<String> initializedLazyFields = new HashSet<>();
+				for ( int i = 0; i < types.length; i++) {
+					if(!types[i].isAssociationType()) {
+						continue;
+					}
+					final String propertyName = propertyNames[i];
+					CascadeStyle cascadeStyle = cascadeStyles[i];
+					if(cascadeStyle != CascadeStyles.NONE) {
+						initializedLazyFields.add(propertyName);
+						continue;
+					}
+					Object propertyValue = persister.getEntityTuplizer().getGetter(i).get(instance);
+					if(propertyValue == null) {
+						if(types[i].isCollectionType()) {
+							needsInterceptor = true;
+							continue;
+						}
+					}
+					else {
+						if(propertyValue instanceof PersistentAttributeInterceptable) {
+							PersistentAttributeInterceptable interceptableProperty = ((PersistentAttributeInterceptable) propertyValue );
+							PersistentAttributeInterceptor propertyInterceptor = interceptableProperty.$$_hibernate_getInterceptor();
+							if(propertyInterceptor == null) {
+								needsInterceptor = true;
+								continue;
+							}
+							else {
+								if(lazyAttributesMetadata.isLazyAttribute(propertyName)) {
+									initializedLazyFields.add(propertyName);
+								}
+							}
+						}
+						if(types[i].isCollectionType()) {
+							if(lazyAttributesMetadata.isLazyAttribute(propertyName)) {
+								initializedLazyFields.add(propertyName);
+							}
+							continue;
+						}
+					}
+				}
+				if (needsInterceptor) {
+					persister.getBytecodeEnhancementMetadata().injectEnhancedEntityAsProxyInterceptor(instance, getEntityKey(), session);
+					interceptor = interceptable.$$_hibernate_getInterceptor();
+					for (String propertyName : initializedLazyFields) {
+						interceptor.attributeInitialized(propertyName);
+					}
+				}
```

Comment:
Copying the [conversation from the original PR](https://github.com/hibernate/hibernate-orm/pull/4339#discussion_r742641143).

From @yrodiere:

> So, here, you're assigning an interceptor to a newly intercepted entity.
> 
> Unless I'm misunderstood, all the code above is trying to isolate `null` associations, so that the interceptor will consider those associations as un-initialized and will try to initialize them the next time we access them.
> 
> Fair enough.
> 
> I think this could be improved, though. This is a lot of code to execute for each property of each newly inserted entity. Maybe it would be a good idea to build a list of "potentially uninitialized properties on persist" on bootstrap, storing it somewhere (e.g. in the entity persister), so that here you only have to go through that list and check for nulls? Then you'll consider non-null properties, as well as all other non-association properties, as initialized.

From @busitech:

> True, but most of this code does not execute if the property is not an association type:
> 
> `if( ! types[i].isAssociationType()) continue;`
> 
> Still, I definitely agree with you, a few iterations can be saved, and I thought about this as I was writing this code.  However, since I am not on the ORM team, I have chosen to avoid creating new class member variables, and changing the signature of existing methods, and just tried to work within the framework.  I wanted the ORM team to decide where and how to optimize, especially since ORM 6.0 is a major release.  I was willing to do 95% of the work to get this up and running in ORM 5.x, but the remaining 5% to fully optimize might double our investment, and then might be thrown out when we see the changes in ORM 6.0?  Ideally, we should try to interrupt the ORM team to look at this while 6.0 is unreleased so they can introduce any needed breaking changes at the major release, rather than at a point release.
> 
> Just as Search is your baby, ORM is someone else's baby, and I was very cognizant of this.  ORM is a huge framework that has been in development for more than 20 years, and after getting into it, it is really a beautiful work of art.  Looking at the code, I saw many loops which iterate over the entity property list, so my impression was that it must not cost enough in terms of performance to require optimization and an increase in memory footprint.
> 
> If we can get the larger issues agreed upon with the ORM team, then we can certainly start to buff and wax the surface.
> 
> It also could be that the ORM team sees the bigger picture right away, and wants to take a different approach entirely, such as solving the issue with a better architectural design, wherein this workaround gets thrown away.  Let's engage them ASAP.